File: python/p078.py
# 
# Solution to Project Euler problem 78
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 

import itertools


MODULUS = 10**6

def compute():
	partitions = [1]
	for i in itertools.count(len(partitions)):
		# We calculate partitions[i] mod 10^6 using a formula based on generalized pentagonal numbers:
		#   partitions(i) =   partitions(i - pentagonal(1)) + partitions(i - pentagonal(-1))
		#                   - partitions(i - pentagonal(2)) - partitions(i - pentagonal(-2))
		#                   + partitions(i - pentagonal(3)) + partitions(i - pentagonal(-3))
		#                   - partitions(i - pentagonal(4)) - partitions(i - pentagonal(-4))
		#                   + ...,
		#   where pentagonal(j) = (3*n^2 - n) / 2, and
		#   we stop the sum when i - pentagonal(+/-j) < 0.
		# Note that for j > 0, pentagonal(j) < pentagonal(-j) < pentagonal(j+1).
		# 
		# (The formula is used without mathematical justification;
		# see https://en.wikipedia.org/wiki/Partition_(number_theory)#Generating_function .)
		item = 0
		for j in itertools.count(1):
			sign = -1 if j % 2 == 0 else +1
			index = (j * j * 3 - j) // 2
			if index > i:
				break
			item += partitions[i - index] * sign
			index += j  # index == (j * j * 3 + j) // 2
			if index > i:
				break
			item += partitions[i - index] * sign
			item %= MODULUS
		
		# Check or memoize the number
		if item == 0:
			return str(i)
		partitions.append(item)


if __name__ == "__main__":
	print(compute())

File: java/p078.java
/* 
 * Solution to Project Euler problem 78
 * Copyright (c) Project Nayuki. All rights reserved.
 * 
 * https://www.nayuki.io/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public final class p078 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p078().run());
	}
	
	
	private static final int MODULUS = Library.pow(10, 6);
	
	public String run() {
		for (int limit = 1; ; limit *= 2) {
			int result = search(limit);
			if (result != -1)
				return Integer.toString(result);
		}
	}
	
	
	private static int search(int limit) {
		/* 
		 * partitions[i] is {the number of ways i can be written
		 * as an unordered sum of positive integers} mod 10^6.
		 * Note that the partition function P(n, k) can be computed with
		 * dynamic programming using only 1 dimension for memoization.
		 */
		int[] partitions = new int[limit];
		partitions[0] = 1;
		for (int i = 1; i < limit; i++) {
			for (int j = i; j < limit; j++)
				partitions[j] = (partitions[j] + partitions[j - i]) % MODULUS;
		}
		for (int i = 0; i < limit; i++) {
			if (partitions[i] == 0)
				return i;
		}
		return -1;
	}
	
}
